From 867af1854490040617d6b2ee81c5dc9feef994cf Mon Sep 17 00:00:00 2001
From: Junping Du <junping_du@apache.org>
Date: Thu, 6 Aug 2015 07:00:42 -0700
Subject: [PATCH 1292/2848] MAPREDUCE-6443. Add JvmPauseMonitor to
 JobHistoryServer. Contributed by Robert Kanter.

(cherry picked from commit c8469357bad481ca8e341050553a5ae6d98bf8e5)

Conflicts:
	hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/JobHistoryServer.java

Change-Id: If324f2669219d6584682cb32cd37a8e2f06f6239
---
 .../hadoop/mapreduce/v2/hs/JobHistoryServer.java   |   14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/JobHistoryServer.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/JobHistoryServer.java
index 75b7c5b..aa0506d 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/JobHistoryServer.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/JobHistoryServer.java
@@ -37,6 +37,7 @@
 import org.apache.hadoop.service.AbstractService;
 import org.apache.hadoop.service.CompositeService;
 import org.apache.hadoop.util.ExitUtil;
+import org.apache.hadoop.util.JvmPauseMonitor;
 import org.apache.hadoop.util.ShutdownHookManager;
 import org.apache.hadoop.util.StringUtils;
 import org.apache.hadoop.yarn.YarnUncaughtExceptionHandler;
@@ -70,6 +71,7 @@
   private HSAdminServer hsAdminServer;
   private HistoryServerStateStoreService stateStore;
   private KilledHistoryService killedHistoryService;
+  private JvmPauseMonitor pauseMonitor;
 
   // utility class to start and stop secret manager as part of service
   // framework and implement state recovery for secret manager on startup
@@ -142,6 +144,12 @@ protected void serviceInit(Configuration conf) throws Exception {
     addService(aggLogDelService);
     addService(hsAdminServer);
     addService(killedHistoryService);
+
+    DefaultMetricsSystem.initialize("JobHistoryServer");
+    JvmMetrics jm = JvmMetrics.initSingleton("JobHistoryServer", null);
+    pauseMonitor = new JvmPauseMonitor(getConfig());
+    jm.setPauseMonitor(pauseMonitor);
+
     super.serviceInit(config);
   }
 
@@ -192,14 +200,16 @@ public static InetSocketAddress getBindAddress(Configuration conf) {
 
   @Override
   protected void serviceStart() throws Exception {
-    DefaultMetricsSystem.initialize("JobHistoryServer");
-    JvmMetrics.initSingleton("JobHistoryServer", null);
+    pauseMonitor.start();
     super.serviceStart();
   }
   
   @Override
   protected void serviceStop() throws Exception {
     DefaultMetricsSystem.shutdown();
+    if (pauseMonitor != null) {
+      pauseMonitor.stop();
+    }
     super.serviceStop();
   }
 
-- 
1.7.9.5

